contracts/bonding/BondingManager.sol

```diff
@@ -350,142 +309,231 @@ contract BondingManager is IBondingManager, Controllable {
+    function updateTranscoderFeePool(
+        address _transcoder,
+        uint256 _fees,
+        uint256 _claimBlock,
+        uint256 _transcoderTotalStake
+    )
+        external
+        whenSystemNotPaused
+        onlyJobsManager
+        returns (bool)
+    {
+        Transcoder storage t = transcoders[_transcoder];
+
+        // Transcoder must be valid
+        require(transcoderStatus(_transcoder) == TranscoderStatus.Registered);
+
+        uint256 currentRound = roundsManager().currentRound();
+        uint256 delegatorsFeeShare = _fees.mul(t.feeShare).div(100);
+
+        t.tokenPoolsPerRound[currentRound].feePool.push(ClaimFees({
+            claimBlock: _claimBlock,
+            fees: delegatorsFeeShare,
+            transcoderTotalStake: _transcoderTotalStake
+        }));
+
+        // Calculate transcoder fees including share of delegator fees based upon the amount it bonded to self
+        uint256 transcoderFeeShare = _fees.sub(delegatorsFeeShare).add(delegatorsFeeShare.mul(t.bondedAmount).div(transcoderTotalStake(_transcoder)));
+        // Update transcoder stake with fee share
+        t.bondedAmount = t.bondedAmount.add(transcoderFeeShare);
+        // Update transcoder total bonded stake with fee share
+        transcoderPools.increaseTranscoderStake(_transcoder, _fees);
+
+        return true;
     }
 
     /*
-     * @dev Check if transcoder is calling reward in correct range of blocks for its time window
-     * @param _transcoder Address of transcoder
+     * @dev Slash a transcoder. Slashing can be invoked by the protocol or a finder.
+     * @param _transcoder Transcoder address
+     * @param _finder Finder that proved a transcoder violated a slashing condition. Null address if there is no finder
+     * @param _slashAmount Percentage of transcoder bond to be slashed
+     * @param _finderFee Percentage of penalty awarded to finder. Zero if there is no finder
      */
-    function validRewardCall(address _transcoder) public constant returns (bool) {
-        // Check if transcoder is in active set
-        if (!isActiveTranscoder[_transcoder]) return false;
-        // Check if already called for current cycle
-        if (transcoders[_transcoder].rewardRound == roundsManager().currentRound() && transcoders[_transcoder].rewardCycle == roundsManager().cycleNum()) return false;
+    function slashTranscoder(
+        address _transcoder,
+        address _finder,
+        uint64 _slashAmount,
+        uint64 _finderFee
+    )
+        external
+        whenSystemNotPaused
+        onlyJobsManager
+        returns (bool)
+    {
+        // Transcoder must be valid
+        require(transcoderStatus(_transcoder) == TranscoderStatus.Registered);
+
+        Transcoder storage t = transcoders[_transcoder];
+
+        uint256 penalty = t.bondedAmount.mul(_slashAmount).div(100);
+
+        if (penalty > t.bondedAmount) {
```

Comment:
Emptying out a transcoder's bond if the penalty is greater than its current bond amount